src/authorization/access_control.rs

```diff
@@ -0,0 +1,209 @@
+    }
+
+    pub fn set(&mut self, status: BTreeMap<AccessType, bool>) {
+        self.status = status;
+    }
+
+    pub fn is_allowed(&self, access: AccessType) -> bool {
+        match self.status.get(&access) {
+            Some(true) => true,
+            _ => false,
+        }
+    }
+}
+
+impl PublicUserAccess {
+    pub fn new(status: BTreeMap<AccessType, bool>) -> Self {
+        PublicUserAccess { status }
+    }
+
+    pub fn set(&mut self, status: BTreeMap<AccessType, bool>) {
+        self.status = status; // todo: filter out Queries
+    }
+
+    /// Returns `Some(true)` if `access` is allowed and `Some(false)` if it's not.
+    /// `None` means that `User::Anyone` permissions apply.
+    pub fn is_allowed(&self, access: AccessType) -> Option<bool> {
+        match access {
+            AccessType::Read => Some(true), // It's Public data, so it's always allowed to read it.
+            _ => match self.status.get(&access) {
+                Some(true) => Some(true),
+                Some(false) => Some(false),
+                None => None,
+            },
+        }
+    }
+}
+
+pub trait AccessListTrait: Clone + Eq + Ord + Hash + Serialize + DeserializeOwned {
+    fn is_allowed(&self, user: &PublicKey, access: AccessType) -> bool;
+    fn expected_data_version(&self) -> u64;
+    fn expected_owners_version(&self) -> u64;
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub struct PrivateAccessList {
+    pub access_list: BTreeMap<PublicKey, PrivateUserAccess>,
+    /// The expected index of the data at the time this grant status change is to become valid.
+    pub expected_data_version: u64,
+    /// The expected index of the owners at the time this grant status is to become valid.
+    pub expected_owners_version: u64,
+}
+
+impl PrivateAccessList {
+    pub fn access_list(&self) -> &BTreeMap<PublicKey, PrivateUserAccess> {
+        &self.access_list
+    }
+}
+
+impl AccessListTrait for PrivateAccessList {
+    fn is_allowed(&self, user: &PublicKey, access: AccessType) -> bool {
+        match self.access_list.get(user) {
+            Some(access_status) => access_status.clone().is_allowed(access),
+            None => false,
+        }
+    }
+
+    fn expected_data_version(&self) -> u64 {
+        self.expected_data_version
+    }
+
+    fn expected_owners_version(&self) -> u64 {
+        self.expected_owners_version
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub struct PublicAccessList {
+    pub access_list: BTreeMap<User, PublicUserAccess>,
+    /// The expected index of the data at the time this grant status change is to become valid.
+    pub expected_data_version: u64,
+    /// The expected index of the owners at the time this grant status change is to become valid.
+    pub expected_owners_version: u64,
+}
+
+impl PublicAccessList {
+    fn is_allowed_(&self, user: &User, access: AccessType) -> Option<bool> {
```

Comment:
Mm I agree it's an iffy naming practice. Could need some change, I don't remember quite now why it was done like this. Will look it up.

Reply:
The problem is the lack of method overloading in rust.
The one with underscore is a private one that is called for a more specific case.
I tried putting all validation in the trait impl, but that looked very messy.
I think that underscore is the lesser bad of the two.
But some distinguishing word could be added of course.

Reply:
Yes, adding a word expressing the more specific case would be good. I don't seem to have seen such naming otherwise in rust.

Reply:
I was actually wrong, the private one is the less specific.
The hard part though is that there is no difference in semantics so the extra word will be very artificial.

I don't know how you usually do, but adding the types of the parameters in the function name is not a very elegant style. The extension of that is to not name things for what they represent, i.e. intent, (user for PublicKey) but always for what they are (implementation detail).

The public one checks if the user (represented as public key) access combo is allowed.
The private one checks if the user (represented by enum User) access combo is allowed.

Reply:
If there is really no need for having different name (not 100% convinced, is_allowed is so very generic seem we could be a bit more specific, but also the function seem really simple so maybe it is fine (Yes, adding type parameter is not a great idea in general)).

I think you need not use `_` though:
https://doc.rust-lang.org/book/ch19-03-advanced-traits.html

> Fully Qualified Syntax for Disambiguation: Calling Methods with the Same Name
Nothing in Rust prevents a trait from having a method with the same name as another trait’s method, nor does Rust prevent you from implementing both traits on one type. It’s also possible to implement a method directly on the type with the same name as methods from traits.

Reply:
Yep, I saw this as well. Just seemed like a lot of bloat with a couple of traits + impl, when an underscore solves it. But, it is more correct of course. 
Not sure here, because I'm not a fan of that underscore, but sometimes I think terse and slim wins over "correct".